runtime/executor/src/main/java/org/apache/nemo/runtime/executor/datatransfer/OutputCollectorImpl.java

```diff
@@ -15,119 +15,85 @@
  */
 package org.apache.nemo.runtime.executor.datatransfer;
 
-import org.apache.nemo.common.Pair;
 import org.apache.nemo.common.ir.OutputCollector;
+import org.apache.nemo.common.ir.vertex.IRVertex;
+import org.apache.nemo.common.ir.vertex.OperatorVertex;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.*;
 
 /**
  * OutputCollector implementation.
+ * This emits four types of outputs
+ * 1) internal main outputs: this output becomes the input of internal Transforms
+ * 2) internal additional outputs: this additional output becomes the input of internal Transforms
+ * 3) external main outputs: this external output is emitted to OutputWriter
+ * 4) external additional outputs: this external output is emitted to OutputWriter
  *
  * @param <O> output type.
  */
 public final class OutputCollectorImpl<O> implements OutputCollector<O> {
   private static final Logger LOG = LoggerFactory.getLogger(OutputCollectorImpl.class.getName());
-  private final Set<String> mainTagOutputChildren;
-  // Use ArrayList (not Queue) to allow 'null' values
-  private final ArrayList<O> mainTagElements;
-  // Key: Pair of tag and destination vertex id
-  // Value: data elements which will be input to the tagged destination vertex
-  private final Map<Pair<String, String>, ArrayList<Object>> additionalTaggedChildToElementsMap;
+
+  private final IRVertex irVertex;
+  private final Set<OperatorVertex> internalMainOutputs;
```

Comment:
For this variable, and the following 3 variables, change the types

Set-OperatorVertex -> ArrayList-OperatorVertex
Set-OutputWriter -> ArrayList-OutputWriter

since lists are faster to iterate?